### PR TITLE
Feature/minimum number of instances

### LINKED ISSUE #1033: Require minimum number of instances
I want the ability to require a minimum number of instances of a service. 

Currently it can see if a service is up and running or not but fails to see a problem if someone would turn scale down to 0 or delete the entire kubernetes workload, it would simply disappear from the Application list and Wallboard with no error.

I could turn off auto-deregistration but that would require manual purging of services after every image version update since k8s simply redeloys and deletes the services with new names and everything.

If there is a way to do this currently I've failed to find it and I'm sorry in advance.

### PR DESCRIPTION
**Pull request by Glosur**
Closes #1033 

Here is a shoot to implement that. I took what seemed the best way for me. What do you think ?

I'll try to implement more java tests but i'd like to have a first 'review' beforehand.

### CODE REVIEW

File: spring-boot-admin-server/src/main/java/de/codecentric/boot/admin/server/web/ApplicationsController.java [hunk 1]
```diff
@@ -60,18 +63,24 @@
...
 
     @GetMapping(path = "/applications", produces = MediaType.APPLICATION_JSON_VALUE)
     public Flux<Application> applications() {
         return registry.getInstances()
                        .filter(Instance::isRegistered)
                        .groupBy(instance -> instance.getRegistration().getName())
-                       .flatMap(grouped -> toApplication(grouped.key(), grouped));
+                       .flatMap(grouped -> toApplication(grouped.key(), grouped))
+                       .concatWith(Flux.fromIterable(requiredApplications.values()))
+                       .distinct((app) -> app.getName());
```
Comment (Glosur) [inline spring-boot-admin-server/src/main/java/de/codecentric/boot/admin/server/web/ApplicationsController.java]: Here, I had no idea how to choose the 'required' application only if no other where present

**Comment by joshiste** [comment]
I'm currently thinking about a more elaborate solution, which wouldn't need to configure the names in the sba server...

... The solution would include a derived application read model and also allow to throw new events (on application basis) so that we could do notifications based on the application and not the instances ...

**Comment by Glosur** [comment]
Ok :)

Are you going to implement it, or can i give it a try (given some advices ?)

**Comment by Glosur** [comment]
Also, at some point, wouldn't you need to know all the existing applications ?

In my use case, we have some test environments where we only have one instance of each microservice. So when we restart eureka/SBA, we need to manually check if all instances are up, because the ones that are down never appeared

**Comment by joshiste** [comment]
> Also, at some point, wouldn't you need to know all the existing applications ?... because the ones that are down never appeared

Unless you make the data for SBA persistent, yes

**Comment by Glosur** [comment]
I'll try to implment what you said, with an `ApplicationRepository `getting fulfilled as instances come up, and a property to toggle keeping OFFLINE applications
I'll update applications listening to the `InstanceEvent`s.
It will allow me to simplify the `ApplicationController`

Does that sound good to you @joshiste ?

**Comment by joshiste** [comment]
sounds like a plan. And yes the applications-controller should get "dumb" at the end.

**Comment by dakotamaker** [comment]
Hi, @Glosur have you made any progress with this PR at all by chance?

**Comment by Glosur** [comment]
Hi, @dakotamaker 

I'm slowly advancing. I'll try to submit a first shot this week end.

**Comment by joshiste** [comment]
any update?

**Comment by joshiste** [comment]
closed due to inactivity

**Pull request closed without merging**